src/Auth0.OidcClient.Core/Tokens/IdTokenValidator.cs

```diff
@@ -0,0 +1,151 @@
+
+            // Audience
+            var audienceCount = token.Audiences.Count();
+            if (audienceCount == 0)
+                throw new IdTokenValidationException("Audience (aud) claim must be a string or array of strings present in the ID token.");
+            if (!token.Audiences.Contains(required.Audience))
+                throw new IdTokenValidationException($"Audience (aud) claim mismatch in the ID token; expected \"{required.Audience}\" but was not one of \"{String.Join(", ", token.Audiences)}\".");
+
+            // Expires at
+            var exp = GetEpoch(token.Claims, JwtRegisteredClaimNames.Exp);
+            if (exp == null)
+                throw new IdTokenValidationException("Expiration Time (exp) claim must be an integer present in the ID token.");
+            var expiration = exp + required.Leeway.TotalSeconds;
+            if (epochNow >= expiration)
+                throw new IdTokenValidationException($"Expiration Time (exp) claim error in the ID token; current time ({epochNow}) is after expiration time ({expiration}).");
+
+            // Issued at
+            var iat = GetEpoch(token.Claims, JwtRegisteredClaimNames.Iat);
+            if (iat == null)
+                throw new IdTokenValidationException("Issued At (iat) claim must be an integer present in the ID token.");
+            var issued = iat - required.Leeway.TotalSeconds;
+            if (epochNow < issued)
+                throw new IdTokenValidationException($"Issued At (iat) claim error in the ID token; current time ({epochNow}) is before issued at time ({expiration}).");
+
+            // Nonce
+            if (required.Nonce != null)
+            {
+                if (string.IsNullOrWhiteSpace(token.Payload.Nonce))
+                    throw new IdTokenValidationException("Nonce (nonce) claim must be a string present in the ID token.");
+                if (token.Payload.Nonce != required.Nonce)
+                    throw new IdTokenValidationException($"Nonce (nonce) claim mismatch in the ID token; expected \"{required.Nonce}\", found \"{token.Payload.Nonce}\".");
+            }
+
+            // Authorized Party
+            if (audienceCount > 1)
+            {
+                if (string.IsNullOrWhiteSpace(token.Payload.Azp))
+                    throw new IdTokenValidationException("Authorized Party (azp) claim must be a string present in the ID token when Audiences (aud) claim has multiple values.");
+                if (token.Payload.Azp != required.Audience)
+                    throw new IdTokenValidationException($"Authorized Party (azp) claim mismatch in the ID token; expected \"{required.Audience}\", found \"{token.Payload.Azp}\".");
+            }
+
+            // Authentication time
+            if (required.MaxAge.HasValue)
+            {
+                var authTime = GetEpoch(token.Claims, JwtRegisteredClaimNames.AuthTime);
+                if (!authTime.HasValue)
```

Comment:
for the rest of the time checks you use `if (myVar == null)` but here you use `HasValue`. Is there a special reason to use a different assertion here?

Reply:
The rest of them are reference types which can be null.  `TimeSpan` here is a value type so it uses `Nullable<TimeSpan>` and you can use `HasValue` which I think is easier to read.  You can also use `== null` but the compiler basically just maps that back to `HasValue` when it's compiling.